sklearn/pipeline.py

```diff
@@ -869,6 +877,11 @@ def __init__(
         self.verbose = verbose
         self._validate_transformers()
 
+    @property
+    def named_transformers_(self):
```

Comment:
I hadn't realized that. Would something like this be more conformant?

```python
    @property
    def named_transformers_(self):
        try:
            check_is_fitted(self)
        except NotFittedError:
            return Bunch(**dict(self.transformer_list))
        return Bunch(**dict(self.transformer_list_))

    def _update_transformer_list(self, transformers):
        transformers = iter(transformers)
        # use a fit attrib, and reference this post-fit
        self.transformer_list_ = [(name, old if old == 'drop'
                                     else next(transformers))
                                    for name, old in self.transformer_list]

```

Returning something when unfitted might also be technically incorrect, however.

Reply:
@thomasjpfan do you think that it could be better to have a `named_transformers` with the same semantic than in `Pipeline`? Our API is broken there and I am not sure that using the `_` is actually a good idea then.

Reply:
> do you think that it could be better to have a named_transformers with the same semantic than in Pipeline? 

I would be okay with a `named_transformers` to be consistent with `Pipeline`'s `named_steps`:

https://github.com/scikit-learn/scikit-learn/blob/9580530071de2bb8de3410967884c1a7ce6c3375/sklearn/pipeline.py#L247-L249

Reply:
Agreed. @crflynn Would you mind make the change?

Reply:
I've updated to reflect this. Tangentially, any plan to correct the API on Pipeline/FeatureUnion?

Reply:
> Tangentially, any plan to correct the API on Pipeline/FeatureUnion?

The work was started here: https://github.com/scikit-learn/scikit-learn/pull/8350

(I also remember speaking with @glemaitre about this in 2019 😅 )

Reply:
Yep, it is even older than 2019. The difficulties in correcting the API come that we might have allowed some use-cases that the correct API would not allow. So we kind of need to think about all of them and provide the correct API but as well support of all previous use-cases.